projects/ngx-sub-form/src/lib/deprecated/ngx-sub-form.component.ts

```diff
@@ -287,28 +298,45 @@ export abstract class NgxSubFormComponent<ControlInterface, FormInterface = Cont
 
   private handleFormArrayControls(obj: any) {
     Object.entries(obj).forEach(([key, value]) => {
-      if (this.formGroup.get(key) instanceof UntypedFormArray && Array.isArray(value)) {
+      if (
+        this.formGroup.get(key) instanceof UntypedFormArray &&
+        Array.isArray(value)
+      ) {
         const formArray: UntypedFormArray = this.formGroup.get(key) as UntypedFormArray;
-
         // instead of creating a new array every time and push a new FormControl
         // we just remove or add what is necessary so that:
         // - it is as efficient as possible and do not create unnecessary FormControl every time
         // - validators are not destroyed/created again and eventually fire again for no reason
-        while (formArray.length > value.length) {
-          formArray.removeAt(formArray.length - 1);
-        }
-
-        for (let i = formArray.length; i < value.length; i++) {
-          if (this.formIsFormWithArrayControls()) {
-            formArray.insert(i, this.createFormArrayControl(key as ArrayPropertyKey<FormInterface>, value[i]));
-          } else {
-            formArray.insert(i, new UntypedFormControl(value[i]));
-          }
-        }
+        this.removeUnnecassaryObjects(formArray, value);
+        this.addAdditionalObjects(formArray, value, key);
       }
     });
   }
 
+  private addAdditionalObjects(
+    formArray: UntypedFormArray,
+    value: Array<any>,
+    key: string
+  ) {
+    for (let i = formArray.length; i < value.length; i++) {
+      const control = this.formIsFormWithArrayControls()
+        ? this.createFormArrayControl(
+          key as ArrayPropertyKey<FormInterface>,
+          value[i]
+        )
+        : new UntypedFormControl(value[i]);
+      formArray.insert(i, control, { emitEvent: this.emitInitialValueOnInit });
+    }
+  }
+
+  private removeUnnecassaryObjects(formArray: UntypedFormArray, value: Array<any>) {
+    while (formArray.length > value.length) {
+      formArray.removeAt(formArray.length - 1, {
+        emitEvent: this.emitInitialValueOnInit,
+      });
+    }
+  }
```

Comment:
I remember seeing an update on Angular side to clear the array in a simpler way but didn't bring it so that it'd remain compatible with older versions. But it was a **while ago** so I'd say it's safe to switch to it: `formArray.clear()`

Here's a screenshot of what the doc says when hovering the method in the IDE

![image](https://user-images.githubusercontent.com/4950209/175884532-01177670-4551-4475-b594-d449982786bd.png)

I think this will do 2 things:
- Avoid having the `removeUnnecassaryObjects` methods, now calling `formArray.clear()` will be simple enough
- Here if the array has 100 values, you'd be broadcasting the event 100 times. With `formArray.clear` it'd only broadcast the event once when all the values have been removed from the array

Reply:
Okay, I take a look on it

Reply:
Some investigations on that. Yes we should definitely do clear on the formarray. That removes all controls from the array and after that with addAdditionalObjects the controls based on the value array are created new. For each value we add to the formarray we should emit. This not really optimal but this solve a lot of other issues. First the validation of the formarray is calculated new and secondly if another order of the value array is patch this new sorting is patched into the form array. Which is currently not the case. We had this problem today at our project.

Reply:
I changed the old and new implementation to user formarray clear.